sprint_3 гайдай а.ф. 18а когорта

dreamQA opened the pull request:
сделано задание

Commit:
сделано задание

ShamilovR commented on conftest.py:
```diff
@@ -0,0 +1,35 @@
+from selenium import webdriver
+
+
+@pytest.fixture
+    # Генерируем случайное "имя"
+def generate_random_name():
+    name_length = random.randint(5, 10)
+    name = ''.join(random.choices(string.ascii_letters, k=name_length)).capitalize()
+    return name
+
+@pytest.fixture
+    # Генерируем случайный "email"
+def generate_random_email():
+    username_length = random.randint(6, 16)
+    username = ''.join(random.choices(string.ascii_lowercase + string.digits , k=username_length))
+    return f"{username}@yandex.ru"
+
+@pytest.fixture
+    # Гененрируем случаный пароль
+def generate_random_password():
+    password_length = 8
+    characters = string.ascii_letters + string.digits + string.punctuation
+    password = ''.join(random.choices(characters, k=password_length))
```
Можно лучше: фикстуры предназначены для выполнения предусловий тестовых методов, т.е. более сложной логики (инициализация драйвера, открытие необходимой страницы, вызовы апи методов). Генерацию случайных строк следует хранить в вспомогательных модулях (helpers).

ShamilovR commented on locators.py:
```diff
@@ -0,0 +1,61 @@
+# Локаторы для тестов
+
+# Поле для ввода имени на странице регистрации
+name_input = "//*[@id=\"root\"]/div/main/div/form/fieldset[1]/div/div/input" # Поле "имя"
```
Нужно исправить: здесь и далее: использование в локаторе порядкового номера элемента в dom, поиск от html, root, main и абсолютного пути порождает нестабильность локаторов (необходимо привязываться к атрибутам конкретного элемента)

dreamQA replied:
постарался исправить,кроме 1 локатора,оставил комментарий в коде

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,19 @@
+from conftest import *
+
+# Тест успешного выхода из личного кабинета по кнопку "Выход"
+def test_deauthorize_from_personal_account(driver):
```
Нужно исправить: здесь и далее: тестовые сценарии должны быть оформлены в тестовых классах. Такая реализация позволяет объединять несколько сценариев в сьюты в рамках одного модуля.

dreamQA replied:
исправил

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,19 @@
+from selenium.webdriver.common.by import By
+import time
+from locators import *
+from conftest import *
```
Нужно исправить: здесь и далее: фикстуры не нужно импортировать. Если они лежат в conftest то pytest их видит без импорта

dreamQA replied:
исправил

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,19 @@
+
+# Тест успешного выхода из личного кабинета по кнопку "Выход"
+def test_deauthorize_from_personal_account(driver):
+    driver.get("https://stellarburgers.nomoreparties.site/login")
```
Можно лучше: здесь и далее: урлы стоит хранить в отдельном модуле и конкатенировать с базовым, так будет проще поддерживать несколько стендов

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,19 @@
+# Тест успешного выхода из личного кабинета по кнопку "Выход"
+def test_deauthorize_from_personal_account(driver):
+    driver.get("https://stellarburgers.nomoreparties.site/login")
+    driver.find_element(By.XPATH, login_email_input).send_keys("alexandergajdaj18a111@yandex.ru")
```
Можно лучше: здесь и далее: тестовые данные стоит хранить отдельно в data модуле, так поддерживать и переиспользовать будет проще

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,19 @@
+    driver.find_element(By.XPATH, login_button_authorization).click()
+
+    driver.find_element(By.XPATH, personal_account_button).click()
+    time.sleep(1)
```
Нужно исправить: здесь и далее: sleep в тестах плохая практика, т.к. для загрузки страницы/элемента может потребоваться меньше или больше времени. Для ожиданий нужно использовать модуль expected_condition.

dreamQA replied:
исправил

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,19 @@
+    driver.find_element(By.XPATH, deauthorize_button).click()
+
+    time.sleep(5)
+    driver.quit()
```
Нужно исправить: здесь и далее: инициализация/закрытие драйвера должна происходить в рамках фикстур. Такая реализация позволяет избежать дублирования кода.

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,19 @@
+    driver.find_element(By.XPATH, personal_account_button).click()
+    time.sleep(1)
+    driver.find_element(By.XPATH, deauthorize_button).click()
+
```
Нужно исправить: здесь и далее: тесовый сценарий должен заканчиваться проверкой (assert`ом)

dreamQA replied:
исправил

Commit:
1. переписаны локаторы через атрибуты элементов
2. удалены импорты фикстур из тестов
3. тестовые сценарии объедены в классы
4. добавлен модуль expected_condition и убраны все time.sleep
5. добавлены проверки (assert) после каждого теста

ShamilovR commented on tests/tests_navigation.py:
```diff
@@ -0,0 +1,89 @@
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support.ui import WebDriverWait
+from selenium.webdriver.support import expected_conditions as EC
+import time
```
Нужно исправить: неиспользуемый импорт модуля.

ShamilovR commented on tests/tests_navigation.py:
```diff
@@ -0,0 +1,89 @@
+            EC.url_changes("https://stellarburgers.nomoreparties.site/login")
+        )
+        current_url = driver.current_url
+        print(f" Текущий URL: {current_url}")
```
Нужно исправить: здесь и далее: для сообщений подобного рода print`ы избыточны. При желании можно использовать логгеры.

ShamilovR commented on tests/tests_navigation.py:
```diff
@@ -0,0 +1,89 @@
+            EC.element_to_be_clickable((By.XPATH, buns_section))
+        )
+
+        assert driver.find_element(By.XPATH, buns_section), "Переход в раздел 'Соусы' удался"
```
Нужно исправить: здесь и далее для тестов переходов по табам: данный элемент появляется в dom с самого начала загрузки страницы. Проверка переключения таба по нему будет недостоверной. Нужно использовать другой способ, например, проверка изменения атрибута class у элемента.

ShamilovR commented on tests/test_deauthorize_from_personal_account.py:
```diff
@@ -0,0 +1,28 @@
+    # Тест успешного выхода из личного кабинета по кнопке "Выход"
+    def test_deauthorize_from_personal_account(self,driver):
+        driver.get("https://stellarburgers.nomoreparties.site/login")
+        driver.find_element(By.XPATH, login_email_input).send_keys("alexandergajdaj18a111@yandex.ru")
```
Можно лучше: здесь и далее: тестовые данные стоит хранить отдельно в data модуле, так поддерживать и переиспользовать будет проще

Commit:
1. Нужно исправить: неиспользуемый импорт модуля. - Исправлено.
2.Нужно исправить: здесь и далее: для сообщений подобного рода print`ы избыточны. При желании можно использовать логгеры. - Исправлено (удалены)
3.Нужно исправить: здесь и далее для тестов переходов по табам: данный элемент появляется в dom с самого начала загрузки страницы. Проверка переключения таба по нему будет недостоверной. Нужно использовать другой способ, например, проверка изменения атрибута class у элемента. - Проверки переходов по табам переписаны с учетом изменения атрибута class у элемента на примере: assert 'text text_type_main-default' in driver.find_element(By.XPATH, fillings_section).get_attribute('class'), "Переход в раздел 'Начинки' удался"
